### PR TITLE
ISPN-5518 Introduce a separate thread pool for async cache operations

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-5518

### CODE REVIEW

File: core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java [hunk 1]
```diff
@@ -35,6 +35,7 @@
 import org.infinispan.util.concurrent.TimeoutException;
 import org.infinispan.util.logging.Log;
 import org.infinispan.util.logging.LogFactory;
+import org.jgroups.protocols.pbcast.STATE_TRANSFER;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java]: Seems this can be removed.

**Comment by wburns** [comment]
Looks fine to me.  The only thing is should we tweak the default thread counts for the remote and async thread pool so we aren't doubling them though?

@danberindei do you know anything about that test failure?  I was unable to reproduce it when I ran the tests.

**Comment by danberindei** [comment]
@wburns I haven't seen the error on my machine either, let's see how the second build goes. About the pool sizes, considering that the remote executor has 200 threads by default, I think it's ok to have 25 threads in each of these pools. I'll just change `TestCacheManagerFactory` to reduce the number of threads used in the test suite.

**Comment by danberindei** [comment]
I had missed some stuff in the parser, it should be ok now. And I'm pretty sure that the failure in the first build wasn't related to these changes.

**Comment by danberindei** [comment]
PR updated to increase the number of transport executor threads back to 6 - the lower number was causing random failures in the map/reduce tests. Should be ok to reduce them again after ISPN-5484 is in.

**Comment by galderz** [comment]
I don't think this PR is complete. [`CacheImpl` is still referring to the `ASYNC_TRANSPORT_EXECUTOR` for the `asyncExecutor` variable](https://github.com/danberindei/infinispan/blob/ISPN-5518/async_operations_thread_pool/core/src/main/java/org/infinispan/cache/impl/CacheImpl.java#L182). That needs to be changed to `ASYNC_OPERATIONS_EXECUTOR`.

**Comment by galderz** [comment]
Also, although I know the reason for doing this, it's a shame that we have to add yet another thread pool to the mix. I hope we can consolidate some of those at some point... :neutral_face:

**Comment by danberindei** [comment]
Oops, I made the CacheImpl change only in the product PR... it should be ok now.

I totally agree we should try to reduce the number of thread pools, not long ago I was railing against Gustavo introducing the `stateTransferExecutor` :) ISPN-5484 does reduce the need for the `transportExecutor` a bit, but it doesn't eliminate it completely.

**Comment by gustavonalle** [comment]
> I was railing against Gustavo introducing the stateTransferExecutor :)

That one was introduced to avoid clogging the transport executor and stop all messages from flowing. 
I believe the motivation here is the same, I'm wondering if both the stateTransferExecutor and this new one could be consolidated?

**Comment by danberindei** [comment]
@gustavonalle I don't think allowing putAsync operations to block state transfer would be a good idea either, although it's true that we're not blocking anything while applying state.

OTOH I'm still hoping we could replace the `stateTransferExecutor` with `putAll` at some point and avoid the need for a thread pool altogether...

File: core/src/test/java/org/infinispan/configuration/XmlFileParsingTest.java [hunk 2]
```diff
@@ -320,21 +320,28 @@ private void assertNamedCacheFile(EmbeddedCacheManager cm, boolean deprecated) {
...
-      assertEquals(TestCacheManagerFactory.MAX_REQ_EXEC_THREADS, remoteCommandThreadPool.maxThreads());
+      assertEquals(TestCacheManagerFactory.REMOTE_EXEC_MAX_THREADS, remoteCommandThreadPool.maxThreads());
+      assertEquals(TestCacheManagerFactory.REMOTE_EXEC_QUEUE_SIZE, remoteCommandThreadPool.queueLength());
       assertEquals(TestCacheManagerFactory.KEEP_ALIVE, remoteCommandThreadPool.keepAlive());
 
       BlockingThreadPoolExecutorFactory stateTransferThreadPool =
             cm.getCacheManagerConfiguration().stateTransferThreadPool().threadPoolFactory();
-      assertEquals(20, stateTransferThreadPool.maxThreads());
-      assertEquals(5, stateTransferThreadPool.queueLength());
-      assertEquals(60000, stateTransferThreadPool.keepAlive());
```
Comment (wburns) [inline core/src/test/java/org/infinispan/configuration/XmlFileParsingTest.java]: Should this be replaced with ```assertEquals(TestCacheManagerFactory.KEEP_ALIVE, stateTransferThreadPool.keepAlive());```

**Comment by wburns** [comment]
Pulling...

**Comment by wburns** [comment]
Integrated into master, thanks @danberindei !

**Pull request closed without merging**